File: src/ports/outbound/investigation-context.ts
import type { DatadogEventSearchPort } from "./datadog-event-search";
import type { DatadogLogAggregatePort } from "./datadog-log-aggregate";
import type { DatadogLogSearchPort } from "./datadog-log-search";
import type { DatadogMetricCatalogPort } from "./datadog-metric-catalog";
import type { DatadogMetricQueryPort } from "./datadog-metric-query";
import type { GithubSearchPort } from "./github-search";

export interface InvestigationResources {
	logAggregatePort: DatadogLogAggregatePort;
	logSearchPort: DatadogLogSearchPort;
	metricCatalogPort: DatadogMetricCatalogPort;
	metricQueryPort: DatadogMetricQueryPort;
	eventSearchPort: DatadogEventSearchPort;
	datadogSite: string;
	githubScopeOrg: string;
	githubSearchPort: GithubSearchPort;
}

export interface InvestigationRuntime {
	startedAtIso: string;
	channel: string;
	threadTs: string;
	retryCount: number;
}

export interface InvestigationContext {
	resources: InvestigationResources;
	runtime: InvestigationRuntime;
}

File: rust/crates/shared/src/ports/outbound/investigation_context.rs
use std::sync::Arc;

use super::{
    DatadogEventSearchPort, DatadogLogAggregatePort, DatadogLogSearchPort,
    DatadogMetricCatalogPort, DatadogMetricQueryPort, GithubSearchPort, WebSearchPort,
};

pub struct InvestigationResources {
    pub log_aggregate_port: Arc<dyn DatadogLogAggregatePort>,
    pub log_search_port: Arc<dyn DatadogLogSearchPort>,
    pub metric_catalog_port: Arc<dyn DatadogMetricCatalogPort>,
    pub metric_query_port: Arc<dyn DatadogMetricQueryPort>,
    pub event_search_port: Arc<dyn DatadogEventSearchPort>,
    pub datadog_site: String,
    pub github_scope_org: String,
    pub github_search_port: Arc<dyn GithubSearchPort>,
    pub web_search_port: Arc<dyn WebSearchPort>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct InvestigationRuntime {
    pub started_at_iso: String,
    pub channel: String,
    pub thread_ts: String,
    pub retry_count: u32,
}

pub struct InvestigationContext {
    pub resources: InvestigationResources,
    pub runtime: InvestigationRuntime,
}
